### ISSUE TITLE
proxy: twin websocket connections?

### ISSUE THREAD

**Issue by edrex** [mholt/caddy#763]
Companion issue: https://github.com/emersion/net-browserify/issues/7

It appears to be relevant that Caddy opens two websocket connections rather than one.

**Comment by mholt** [comment]
It opens two? How came you by this?

**Comment by edrex** [comment]
well, the backend server logs two connections (see linked issue). Wireshark also shows two HTTP connections to the socket, along with lots of DUP ACKs. I hesitate to post the wireshark dump since

![](https://img.buzzfeed.com/buzzfeed-static/static/2014-10/26/6/enhanced/webdr08/enhanced-14836-1414320930-8.jpg)

**Comment by mholt** [comment]
Haha.

I do not see any information about the backend logging two conncetions -- can you link me to it?

**Comment by edrex** [comment]
I was curious if this was always happening and just some backend servers are tolerant of it, so I tried it with the [loop example](https://gist.github.com/edrex/7af8da336fbd1ec6a2f9) and sure enough, wireshark sees two socket connections.

**Comment by edrex** [comment]
Here is a shot from wireshark only showing http and websocket frames:

![image](https://cloud.githubusercontent.com/assets/14615/14584353/c7ee00ac-03f7-11e6-8dcc-37ae968e9ff9.png)

Frames <= 27 are direct, and frames > 40 are through the proxy. Note the duplicate requests in the second group.

**Comment by edrex** [comment]
@mholt I realized I hadn't posted the log on the linked issue, but I think the evidence above is clearer.

**Comment by klingtnet** [comment]
Maybe this is related to my issue running a [jupyter notebook](http://jupyter.org/) server with caddy acting as a reverse proxy. Websocket connections are working when I access the server directly using SSH portforwarding `ssh -L 12345:localhost:12345 host` but there are sporadic issues when connecting via caddy (sometimes it works, sometimes don't). I haven't figured out the problem yet because `jupyter`s log does not show any error, so I also have to record the traffic with wireshark.

Caddy config:

```ini
jupyter.klingt.net {
	proxy / localhost:10180 {
		proxy_header Host {host}
		proxy_header X-Real-IP {remote}
		proxy_header X-Forwarded-Proto {scheme}
		websocket
	}
}
```

**Comment by edrex** [comment]
@klingtnet yeah, most likely. Stay tuned, fix in progress.

**Comment by mholt** [comment]
@edrex How's it coming? Anything I can help with? Since Go 1.6.2 was released last night, we can release a new version of Caddy to great benefit. Definitely want to make sure this fix gets in there.

**Comment by edrex** [comment]
On the road for a conf. Looked at this for an hour yesterday, made some progress but didn't get it working. Hope to get some hack time today

**Comment by edrex** [comment]
Continuing on #764

**Comment by nemothekid** [comment]
Fixed by #987

**Comment by klingtnet** [comment]
Jupyter notebooks are running fine with caddy 0.9.1!
Thanks to all of you for your efforts to fix this bug!
